File: code/sample_projects/scala/todolist_example/src/main/scala/classes/todolist.scala
package todolist
import scala.collection.mutable.Map

// @Suite?
class ToDoList(
  var author: String,
  var items: Map[Int, ToDoItem]
) {
  var id = 0
  def newID = () =>
    id += 1
    id

  def addItem = (item: ToDoItem) =>
    items(newID()) = item
  
  def removeItemByID = (id: Int) =>
    items -= id
  
  def removeAllItems = () =>
    items.clear()
  
  // def removeDuplicateItems = () =>
  //   var seen = Set.empty[String]
  //   items = items.filter { case (id, item) =>
  //     if (seen.contains(item.content)) {
  //       false
  //     } else {
  //       seen += item.content
  //       true
  //     }
  //   }

  def getAuthor = () =>
    author
  
  def getItems = () =>
    items.values.toList
  
  def getSize = () =>
    items.size
  
  def copy = () =>
    ToDoList(author, items.clone())
  
  override def equals(obj: Any): Boolean = 
    obj match {
      case other: ToDoList =>
        this.author == other.author && this.items == other.items
      case _ => false
    }
}

object ToDoList {
  def empty() = 
    ToDoList("", Map.empty[Int, ToDoItem])
}

case class ToDoItem (
  content: String
)

File: code/sample_projects/python/src/classes/ToDoList.py
class ToDoItem:
    def __init__(self, item: str):
        self.item = item
    
    def __eq__(self, other):
        if not isinstance(other, ToDoItem):
            return NotImplemented
        return self.item == other.item

class ToDoList:
    def __init__(self, author: str):
        self.author = author
        self.items: dict[int, ToDoItem] = {}
        self.id = -1

    def newID(self):
        self.id += 1
        return self.id

    def addItem(self, item: ToDoItem):
        self.items[self.newID()] = item

    def removeItemByID(self, id: int):
        del self.items[id]
    
    def copy(self):
        new_list = ToDoList(self.author)
        for item in self.items.values():
            new_list.addItem(item)
        return new_list

    def __eq__(self, other):
        if not isinstance(other, ToDoList):
            return NotImplemented
        return self.author == other.author and self.items == other.items
    
    def __ne__(self, other):
        if not isinstance(other, ToDoList):
            return NotImplemented
        return not self.__eq__(other)
